lib/posix-process: Start tid numbering from one

flpostolache opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`
 - Platform(s): `kvm`
 - Application(s): `N/A`


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
These commits make so that PID/TID 0 cannot be assigned to any thread. Some library calls can fail when using binary compatibility mode because they assume that this value is reserved by the kernel.

flpostolache committed:
lib/posix-process: Correct MAX_PID config option

This commit changes the tid_map size to reflect the kernel
reservation of the 0 TID/PID value. Before, the first thread
would set the 0 position in all used structures. Now, because
it will be already set, we need to make room for another.

Signed-off-by: Florin Postolache <florin.postolache.of@gmail.com>

flpostolache committed:
lib/posix-process: Reserve PID/TID 0 for kernel

This commit marks the TID/PID as unavailable to mimic the behaviour of
other kernels. This change also make pthread_rwlocks from glibc work in
binary compatibility mode.

Signed-off-by: Florin Postolache <florin.postolache.of@gmail.com>

skuenzer commented on lib/posix-process/process.c:
```diff
@@ -40,6 +40,8 @@
 #include <uk/config.h>
 #include <uk/syscall.h>
 
+#define PIDMAP_SIZE (CONFIG_LIBPOSIX_PROCESS_MAX_PID + 1)
```
Maybe calling it `TIDMAP_SIZE` makes more sense in this source file. I am sorry, we spoke offline about it and the suggestion came from me.

skuenzer commented on lib/posix-process/Config.uk:
```diff
@@ -14,6 +14,8 @@ if LIBPOSIX_PROCESS_PIDS
 		range 1 32768
 		default 32
+		help 
+				This value represents the maximum number of pids that can be created
```
With the changes you did, it is now really the largest PID/TID number that can be created and not the number of PIDs.

flpostolache replied:
By that help string, I meant that, including the start thread, you can create that many threads. With the new changes it is indeed the maximum number that can be created. Because it is easier to understand, I am inclined to completely erase the help message or change it to what you commented. What is your opinion?

skuenzer replied:
Good question, I think because the option says `Largest PID`, we can live without the help text.

skuenzer commented on lib/posix-process/Config.uk:
```diff
@@ -14,6 +14,8 @@ if LIBPOSIX_PROCESS_PIDS
 		int "Largest PID"
 		range 1 32768
 		default 32
```
Maybe set the default to 31 so that the bitmap smoothly fits into 4 `long`s.

skuenzer commented on lib/posix-process/Config.uk:
```diff
@@ -14,6 +14,8 @@ if LIBPOSIX_PROCESS_PIDS
 		int "Largest PID"
 		range 1 32768
```
It may make sense to set the maximum to 32767 in order to fully use the available bitmap in the max case.

skuenzer commented on lib/posix-process/process.c:
```diff
@@ -83,9 +83,10 @@ struct posix_thread {
 
 /**
  * System global lists
+ * NOTE: PID/TID 0 are reserved by the kernel and should never be available
```
```diff
- * NOTE: PID/TID 0 are reserved by the kernel and should never be available
+ * NOTE: We pre-allocate PID/TID 0 which is reserved by the kernel.
+ *       An application should never get PID/TID 0 assigned.
```

skuenzer reviewed (changes requested):
Thanks a lot for this changes. Great!

Can you adopt the commit titles to the following?
`lib/posix-process: Correct MAX_PID config option`
`lib/posix-process: Reserve PID/TID 0 for kernel`

Additionally, a debug message confused me but in the end I think this is correct. If there is no parent, PID `0` is printed at line 283:
```c
uk_pr_debug("Process PID %d created (parent PID: %d)\n",
    (int) pprocess->pid,
    (int) ((pprocess->parent) ? pprocess->parent->pid : 0));
```
However, this is not reflected in the getppid system call (line 532). We should probably also return PID 0 here in the case that there is no parent instead of the own PID. At the moment it makes more sense to me to say this process was directly created by the kernel. Because we reserve PID 0 for the kernel, you can do this change in the same commit as well.

flpostolache committed:
lib/posix-process: Fix debug printing of TIDs

This commit makes so that the correct info will be printed when the current
thread gets killed. Before, instead of the actual TID, the empty TID from
the iterator would be printed.

Signed-off-by: Florin Postolache <florin.postolache.of@gmail.com>

skuenzer reviewed (approved):
Thanks a lot for your work and fixing this!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

andraprs reviewed (approved):
Reviewed-by: Andra Paraschiv <andra@unikraft.org>

Thank you for the updates.
